src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -703,10 +703,8 @@ private void fireCompleted(){
      * @return non-null after the flow has started, even after finished (but may be null temporarily when about to start, or if starting failed)
      */
     public @CheckForNull FlowExecution getExecution() {
-        if (executionLoaded || execution == null) {  // Avoids highly-contended synchronization on run
-            return execution;
-        } else {  // Try to lazy-load execution
-            synchronized (this) {  // Double-checked locking rendered safe by use of volatile field
```

Comment:
I don't have much time to try to reason about the locking here in detail, but for what it's worth, maybe it would be safer to preserve `synchronized (this)` inside of `synchronized(getMetadataGuard())`? It's not clear to me whether `WorkflowRun` itself not being locked here could be problematic.

And was there a particular reason to remove the short-circuit case that skips locking when `executionLoaded`?

Really at this point I am very tempted to delete `metadataGuard` to reduce all of the locking complexity and then see if lock contention really does seem to still be a performance issue these days. Maybe we could add a system property to make `getMetadataGuard()` return `this` and then we could test that for a bit?

Reply:
Tricky to look for lock contention under load. We would have to set up a special test environment for that.